### PR TITLE
Alternative implementation of ParamOverrides

### PR DESCRIPTION
**Pull request by ceball**
I've re-implemented ParamOverrides, following the suggestion made in #85 about sharing parameter objects. This PR isn't meant for merging at this stage, I'd just like some comments on the basic idea.

This new implementation passes all the tests (including one I added to check that dynamic overrides work), although the implementation's not yet complete (just missing some minor methods that the current ParamOverrides has). However, what I'd like to know is: is the whole idea crazy? If it's not crazy, is there a way to make the implementation simpler? I've left extra comments in the code for now to explain what I was thinking.

In case it's easier to look at the class in one piece:
https://github.com/ceball/param/blob/bug85_newParamOverrides/param/parameterized.py#L1580

(Also, note that I haven't considered speed.)

### CODE REVIEW

**Comment by ceball** [comment]
Also, what do you think about this? Say a Parameterized instance ``a`` has a mutable parameter, e.g. ``b = param.List([1,2])``. If someone creates a ParamOverrides and sets that parameter to a new list either during initialization (``ParamOverrides(a,{'b':[3,4]})``) or afterwards (``ParamOverrides(a).b=[3,4]``), no problem. But if someone edits the mutable parameter without first having set it (``ParamOverrides(a).b[0]=3``), they'll actually be editing the parameter on the overridden instance (``a.b`` will be ``[3,2]`` in this case). To me that's probably ok with suitable documentation, but with a more complex implementation it would be possible to do better for ``instantiate=True`` mutable parameters.

(I think the above behavior is the same for the existing ParamOverrides.)

**Comment by jbednar** [comment]
Seems ok to me, when finished.  Is it really necessary to make a new class with all the parameters?  I.e., is there no easy way just to point to the parameters as they are in the overridden class?  I haven't tried either approach, so I don't know.

I don't think the mutable object behavior you describe is worth putting in a special case for; people already have to know to deal with mutable objects differently in general.

**Comment by ceball** [comment]
On Sun, Feb 15, 2015, at 07:23 PM, James A. Bednar wrote:
> Seems ok to me, when finished.  Is it really necessary to make a new
> class with all the parameters?  I.e., is there no easy way just to point
> to the parameters as they are in the overridden class? 

A parameterized instance doesn't have pointers to its parameter objects,
except via its class. The new class is where I store the pointers to the
overridden parameterized instance's parameter objects. No new parameter
objects are created (except the standard parameters of Parameterized,
since ParamOverrides is now a subclass of that).

Is this what you meant, or did you mean could I avoid having a new class
by storing elsewhere the pointers to the parameter objects (i.e. doing
something different from a normal Parameterized)?

Or am I on the wrong track completely?

**Comment by jbednar** [comment]
You're the expert on Param, not me, so I'd usually be the one on any wrong track!  

What I meant was that I don't see why you need a *new* class to store the pointers to the parameterized instance's parameter objects.  Doesn't overridden's class already exist, with precisely the right set of parameter objects?  So I'm imagining (very vaguely, so that's as much detail as I can give) somehow pointing to the existing class, so that we don't have to make a new one.  But there could easily be technical reasons why this is not feasible.

**Comment by philippjfr** [comment]
Is there any chance of finishing this up? The fact that ParamOverrides doesn't actually validate types is a major issue.

**Comment by ceball** [comment]
When resuming this work, please consider addressing or closing:
* #36 
* #134 
* #173

**Comment by ceball** [comment]
I think we've established that ParamOverrides has moved from something we're trying out to see if we want, to something that we use, so it's time to finish this! ;) I'll bring this PR up to date.

@jbednar: 
> What I meant was that I don't see why you need a new class to store the pointers to the parameterized instance's parameter objects. Doesn't overridden's class already exist, with precisely the right set of parameter objects? So I'm imagining (very vaguely, so that's as much detail as I can give) somehow pointing to the existing class, so that we don't have to make a new one. But there could easily be technical reasons why this is not feasible.

Yes, the overridden's class already exists, and yes it has exactly the right set of parameter objects. But...

Parameter objects are stored on the class. Different instances of ParamOverrides will override the parameters of different parameterizeds and will hence need different parameters. 

Your question is about having no new class, but in that case the ParamOverrides class would end up with links to parameters on multiple different parameterized classes i.e. it would not have the right parameters for any one parameterized class. 

My implementation is one new class for each ParamOverrides instance created, to store the links to the overridden's parameters. I did that because it's easy to do. Each new class is a subclass of overridden's class. However, it would instead be possible to create only one new class per *type* of overridden. I don't see how it's possible to do it with fewer classes than that while keeping the implementation of ParamOverrides inheriting from Parameterized. But to have only one new class per type of overridden, we'd need to keep the classes around.

Unless I missed something? Quite possible!

**Comment by ceball** [comment]
What's left?

* [ ] add missing methods noted in code TODO
* [ ] test with holoviews
* [ ] test with topographica?
* [ ] test performance?
* [ ] fixed #134?
* [ ] fixed #173?
* [ ] address #36?

**Comment by jlstevens** [comment]
I'm very happy to see these improvements to ParamOverride but I'm not sure about the decorator and corresponding underscore attribute described in #36.

**Comment by ceball** [comment]
@jbednar: Maybe you could review this again? I think we could merge right after the next param release and then see what happens for a while. I've tested with holoviews but not yet topographica (the two things most likely to exercise paramoverrides the most). 

Notes:

* ParamOverrides is less like a dict than before (when it was a dict!). I've added what I think is the minimum dict-like behavior to support holoviews. We could make it emulate dict completely, though.

* ParamOverrides is at the same time slightly more dict-like than before, because now it raises KeyError rather than AttributeError when a name's not found via dict-like access.

* Because ParamOverrides now inherits from Parameterized, the name parameter will effectively always be overridden i.e. in typical usage ``p.name`` will not be the name of the overridden object but of the ParamOverrides object. That could be confusing. 

* Keep the overloads decorator or not (see #36)? If keeping, what about its name (and the name of "_overrides")?

* I haven't tested performance :( The following comment used to exist: ``we'd like __init__ to be fast because it's going to be called a lot``

* Fixes #134 and #173.

**Comment by ceball** [comment]
@jlstevens:
> I'm not sure about the decorator and corresponding underscore attribute described in #36.

Yep, I agree. Wrote my comment above at the same time as yours. After initially liking the way it seems to make things look a bit nicer, I think I prefer the explicit ``p = ParamOverrides(...)``.

**Comment by ceball** [comment]
Another note: #176 (issue #104) should be merged first or at the same time, or else holoviews will have spurious failing tests next time holoviews' travis runs (it takes param from github, I think).

For reference:

```
======================================================================
ERROR: Assert histogram shares colormapper
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cball/code/ioam/holoviews/tests/testplotinstantiation.py", line 839, in test_side_histogram_cmapper
    plot = bokeh_renderer.get_plot(img.hist())
  File "/Users/cball/code/ioam/holoviews/holoviews/core/element.py", line 41, in hist
    individually=individually, dimension=d, **kwargs)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 1881, in __new__
    return inst.__call__(*args,**params)
  File "/Users/cball/code/ioam/holoviews/holoviews/core/operation.py", line 125, in __call__
    self.p = param.ParamOverrides(self, params)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 1758, in __init__
    Parameterized.__init__(self,**dict_)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 1041, in __init__
    self._setup_params(**params)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 978, in override_initialization
    fn(parameterized_instance,*args,**kw)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 1441, in _setup_params
    setattr(self,name,val)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 537, in __set__
    self._check_value(val)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 534, in _check_value
    raise ValueError("String '%s' only takes a string value: %s"%(self._attrib_name,val))
ValueError: String 'dimension' only takes a string value: None

======================================================================
ERROR: test_side_histogram_no_cmapper (tests.testplotinstantiation.TestBokehPlotInstantiation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cball/code/ioam/holoviews/tests/testplotinstantiation.py", line 827, in test_side_histogram_no_cmapper
    plot = bokeh_renderer.get_plot(points.hist())
  File "/Users/cball/code/ioam/holoviews/holoviews/core/element.py", line 41, in hist
    individually=individually, dimension=d, **kwargs)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 1881, in __new__
    return inst.__call__(*args,**params)
  File "/Users/cball/code/ioam/holoviews/holoviews/core/operation.py", line 125, in __call__
    self.p = param.ParamOverrides(self, params)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 1758, in __init__
    Parameterized.__init__(self,**dict_)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 1041, in __init__
    self._setup_params(**params)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 978, in override_initialization
    fn(parameterized_instance,*args,**kw)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 1441, in _setup_params
    setattr(self,name,val)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 537, in __set__
    self._check_value(val)
  File "/Users/cball/code/ceball/param/param/parameterized.py", line 534, in _check_value
    raise ValueError("String '%s' only takes a string value: %s"%(self._attrib_name,val))
ValueError: String 'dimension' only takes a string value: None
```

**Comment by ceball** [comment]
Testing with topographica has revealed two things missing from this implementation vs the previous one: (1) support for pickling, and (2) support for arbitrary (non parameter) attributes on parameterized function instances.

**Comment by ceball** [comment]
Topographica's tests also show a use case I'd completely forgotten about: `p=ParamOverrides(...); f(**p)`. I think `p` has to *be* a dict for that to work (i.e. `p` can't just have the right methods). Or maybe it's just got to be a `Mapping` (abstract base class) - can't remember if that's in python 2.6 or not. Either way I think that makes it difficult for the implementation proposed in this PR to be backwards compatible.

E.g. from imagen:

```python
class GaussianCloud(Composite, TimeAware):
    ...
    def __call__(self,**params_to_override):
        p = ParamOverrides(self,params_to_override)
        ...
        return super(GaussianCloud,self).__call__(**p)
```

**Comment by ceball** [comment]
Also, missing `param_keywords()` method.

**Comment by ceball** [comment]
Summary so far! :(

Currently, we use ParamOverrides like this: `p = ParamOverrides(overridden, overrides)`, where `overridden` is a Parameterized instance and `overrides` is a dict. Typically we use this in the `__call__` method of `overridden`.

I think we want `p` to behave a lot like `overridden`. When you try to set things on `p`, you should get param's usual behaviors (e.g. value checking). When you get things from `p`, you should get param's usual behaviors (e.g. dynamic values). For example, this should work:

```python
class A(param.Parameterized):
    x = param.String(default='hi')
    y = param.Number(default=0.5, bounds=(0,1))

a = A()

try:
    param.ParamOverrides(a, dict(x=1))
except ValueError:
    pass
else:
    raise

param.ParamOverrides(a, dict(y=lambda:0.75)).y == 0.75

try:
    param.ParamOverrides(a, dict(y=lambda:1.5)).y
except ValueError:
    pass
else:
    raise
```

If ParamOverrides doesn't work just like param normally does I think it's confusing. We have at least one example of that already: "ParamOverrides stops Filename search_paths from working" (https://github.com/ioam/param/issues/134), but there'll be others. I've also found myself wanting to use the normal methods available on Parameterized, too. E.g. "ParamOverrides: support get_param_values()" (https://github.com/ioam/param/issues/173). How far should we go? I think ParamOverrides should be just like the overridden instance.

Do you agree with the above? If you think it's too much, and we should instead have ParamOverrides support only a specific, smaller set of things, then please say. 

I thought the simplest and easiest way to achieve the above would be for ParamOverrides to inherit from Parameterized. I felt that alternative suggestions would result either in ParamOverrides being some kind of reimplementation of Parameterized, or in ParamOverrides requiring more magic to access features of Parameterized via overridden than exists in the proposed ParamOverrides-inherits-from-Parameterized implementation.

I created what I think is a bit of an ugly solution, but it works (in that it fixes an important gap in parameter checking, particularly for parameterized functions). I knew I could easily add anything necessary to make the new implementation backwards compatible with the previous one (except for the part about catching parameter errors i.e. the reason for this PR), giving us something working right now (the priority), while hoping in the future we might be able to simplify things/find an entirely better solution. However, it turns out that there's one thing I probably can't do. As noted above (https://github.com/ioam/param/pull/92#issuecomment-328301714), you can only use a Mapping for kwargs. In topographica and imagen (but not holoviews), that appears be done with ParamOverrides a couple of times (e.g. in a `__call__` method: ``p=ParamOverrides(...); ...; super(...).__call__(**p)``. 

So what should I do? Add a method to get a dictionary for use in kwargs and update imagen and topographica and apologize that ParamOverrides may start generating two types of error for your old code (genuine parameter errors, and an error if you try to use it for kwargs without updating your code)? Or maybe make the new ParamOverrides even trickier than it already is by also inheriting from Mapping? Or have I gone down a miserable technical rabbit hole, and can someone suggest an altogether simpler approach?

Jim previously suggested this:

> maybe we should provide a totally different way to do this that's less awkward? E.g. can we offer a method decorator that pushes the current dictionary of parameter values onto a stack, lets the method do what it likes, then pops the original values back off when the method returns?

However, I would be worried about changing "the state of the parameters". You might look at a `__call__` method and decide it doesn't modify anything so you can use it "safely" from multiple threads (or in e.g. a generator), without realizing what's happening to the parameters. It might be particularly surprising for someone to find calling 'a function' with some arguments in one place could affect the function's parameters in another. E.g.

```python
class A(param.ParameterizedFunction):
    x = param.Number(default=0)

                   # pushes __dict__ onto stack and replaces with temp copy into which
    @override_dict # ** params are set; pops __dict__ back off stack on return
    def __call__(self, **params):
        # some long-running operation e.g. a network operation...

a = A.instance(x=0)

with concurrent futures ThreadPool()) as executor: # (I can't quite remember how that goes)
    executor.submit(a) # will likely have a.x = 0
    executor.submit(a, x=10)
    executor.submit(a) # will likely have a.x = 10 because a(x=10) will still be running
```

File: tests/testparameterizedobject.py [hunk 1]
```diff
@@ -297,13 +344,109 @@ def test_missing_key(self):
...
+        
+    def test_param_keywords(self):
+        overrides = ParamOverrides(self.po,{'name':'B','an_extra_thing':1},allow_extra_keywords=True)
+        assert 'an_extra_thing' in overrides.extra_keywords()
+        assert 'an_extra_thing' not in overrides.param_keywords()
+
+    def test_pickle(self):
+        overrides = ParamOverrides(self.po,{'name':'B','an_extra_thing':1},allow_extra_keywords=True)
+        import pickle
+        pkl = pickle.dumps(overrides)
```
Comment (ceball) [inline tests/testparameterizedobject.py]: Should test pickle load also. (Note: pickle load seems to be broken for the current ParamOverrides anyway. Should investigate.)

**Comment by tonyfast** [comment]
Closing this as stale. We are likely to harvest some code from here, but it would require significant work.

**Pull request closed without merging**